File: src/Parrot.Renderers/ForeachRenderer.cs
// -----------------------------------------------------------------------
// <copyright file="ForeachRenderer.cs" company="">
// TODO: Update copyright text.
// </copyright>
// -----------------------------------------------------------------------

namespace Parrot.Renderers
{
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using Parrot.Infrastructure;
    using Parrot.Nodes;
    using Parrot.Renderers.Infrastructure;

    /// <summary>
    /// TODO: Update summary.
    /// </summary>
    public class ForeachRenderer : HtmlRenderer
    {
        public ForeachRenderer(IHost host)
            : base(host)
        {
        }

        public override IEnumerable<string> Elements
        {
            get { yield return "foreach"; }
        }

        public override void Render(IParrotWriter writer, IRendererFactory rendererFactory, Statement statement, IDictionary<string, object> documentHost, object model)
        {
            var localModel = GetLocalModel(documentHost, statement, model);

            //Assert that we're looping over something
            IEnumerable loop = localModel as IEnumerable;
            if (loop == null)
            {
                throw new InvalidCastException("model is not IEnumerable");
            }

            //create a local copy
            IList<object> items = ToList(loop);
            
            //create locals object to handle local values to the method
            Locals locals = new Locals(documentHost);

            for (int i = 0; i < items.Count; i++)
            {
                var item = items[i];

                foreach (var child in statement.Children)
                {
                    var renderer = rendererFactory.GetRenderer(child.Name);
                    locals.Push(IteratorItem(i, items));

                    renderer.Render(writer, rendererFactory, child, documentHost, item);

                    locals.Pop();
                }
            }
        }

        private static object IteratorItem(int index, IList<object> items)
        {
            return new
                {
                    _first = index == 0,
                    _last = index == items.Count - 1,
                    _index = index,
                    _even = index % 2 == 0,
                    _odd = index % 2 == 1
                };
        }

        private IList<object> ToList(IEnumerable loop)
        {
            var list = new List<object>();
            foreach (var item in loop)
            {
                list.Add(item);
            }

            return list;
        }
    }
}

File: src/Parrot.Javascript/scripts/Renderers/foreachRenderer.ts
///<reference path="../Parser/parser.ts" />
///<reference path="./irenderer.ts" />
///<reference path="rendererProvider.ts" />
///<reference path="../Infrastructure/ObjectModelValueProvider.ts" />
///<reference path="../Infrastructure/ValueTypeProvider.ts" />
///<reference path="../Infrastructure/Locals.ts" />
///<reference path="../exceptions.ts" />
///<reference path="./htmlRenderer.ts" />

class ForEachRenderer extends HtmlRenderer {
    defaultChildTag: string = "li";
    rendererProvider: RendererProvider;
    elements: string[] = ["foreach"];
    
    render(statement: Statement, host: any[], model: any, rendererProvider: RendererProvider): string {
        //get the local model
        this.rendererProvider = rendererProvider;

        var localModel = this.getLocalModelValue(host, statement, model);

        if (Object.prototype.toString.call(model) === '[object Array]') {
            var result: string = "";
            for (var i in model) {
                var locals = new Locals(host);
                locals.push(this.iteratorItem(i, model));
                result += this._renderChildren(statement.children, host, model[i], this.defaultChildTag);
                locals.pop();
            }
            return result;
        }

        return "";
    }

    iteratorItem(index: any, items: any[]): any {
        return {
            _first: index == 0,
            _last: index == items.length - 1,
            _index: index,
            _even: index % 2 == 0,
            _odd: index % 2 == 1
        }
    }
}